handler/src/main/java/io/netty/handler/ssl/SslHandler.java

```diff
@@ -989,21 +953,22 @@ public void operationComplete(ChannelFuture future) {
                         }
                     });
                     if (inUnwrap) {
-                        // We may be here because we read data and discovered the remote peer initiated a renegotiation
-                        // and this write is to complete the new handshake. The user may have previously done a
-                        // writeAndFlush which wasn't able to wrap data due to needing the pending handshake, so we
-                        // attempt to wrap application data here if any is pending.
-                        if (status == HandshakeStatus.FINISHED && !pendingUnencryptedWrites.isEmpty()) {
-                            wrap(ctx, true);
-                        }
                         needsFlush = true;
                     }
                     out = null;
                 }
 
+                HandshakeStatus status = result.getHandshakeStatus();
                 switch (status) {
                     case FINISHED:
                         setHandshakeSuccess();
+                        // We may be here because we read data and discovered the remote peer initiated a renegotiation
+                        // and this write is to complete the new handshake. The user may have previously done a
+                        // writeAndFlush which wasn't able to wrap data due to needing the pending handshake, so we
+                        // attempt to wrap application data here if any is pending.
+                        if (inUnwrap && !pendingUnencryptedWrites.isEmpty()) {
```

Comment:
moving to here allows for improved coalescing if the handshake notification generates more writes.